### PR TITLE
Fix is_fundamental template for MSVC

### LINKED ISSUE #30932: Windows Build fails
## 🐛 Bug

For some days, I have a broken build on Windows (cf. appended error message)

## To Reproduce


1. Pull master, submodules etc.
2. (Opt.) clear build directory
3. Build pytorch

## Expected behavior

A finishing build

## Environment

PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A

OS: Microsoft Windows 10 Pro
GCC version: Could not collect
CMake version: version 3.16.0

Python version: 3.7
Is CUDA available: N/A
CUDA runtime version: 10.2.89
GPU models and configuration: GPU 0: Quadro M1200
Nvidia driver version: 441.28
cuDNN version: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\bin\cudnn64_7.dll

Versions of relevant libraries:
[pip3] msgpack-numpy==0.4.4.3

[pip3] numpy==1.17.4+mkl

[pip3] numpydoc==0.9.1

[pip3] torch==1.3.1  # I don't know, what this means - my pip call lists the same development version as conda 

[pip3] torchvision==0.4.2
[conda] blas                      1.0                         mkl  

[conda] libblas                   3.8.0                    14_mkl    conda-forge

[conda] libcblas                  3.8.0                    14_mkl    conda-forge

[conda] liblapack                 3.8.0                    14_mkl    conda-forge

[conda] liblapacke                3.8.0                    14_mkl    conda-forge

[conda] mkl                       2019.4                      245  

[conda] mkl-include               2019.4                      245  

[conda] mkl-service               2.3.0            py37hfa6e2cd_0    conda-forge

[conda] pytorch-sphinx-theme      0.0.24                    dev_0    <develop>

[conda] torch                     1.4.0a0+6e145b4          pypi_0    pypi

[conda] torchvision               0.5.0a0+95131de          pypi_0    pypi


## Additional context

MS Visual Studio got an update lately - maybe the used C++ compiler is worth a check:

`Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\bin\Hostx64\x64\cl.exe`


Error message from build:

```
[95/995] Building CXX object caffe2\CMakeFiles\torch_cpu.dir\core\blob_serialization.cc.obj                                                                                                               
FAILED: caffe2/CMakeFiles/torch_cpu.dir/core/blob_serialization.cc.obj                                                                                                                                    
C:\PROGRA~2\MICROS~2\2019\COMMUN~1\VC\Tools\MSVC\1424~1.283\bin\Hostx64\x64\cl.exe  /nologo /TP -DAT_PARALLEL_OPENMP=1 -DCPUINFO_SUPPORTED_PLATFORM=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DTH_BLAS_MK
L -DWIN32_LEAN_AND_MEAN -D_CRT_SECURE_NO_DEPRECATE=1 -D_OPENMP_NOFORCE_MANIFEST -Dtorch_cpu_EXPORTS -Iaten\src -I..\aten\src -I. -I..\ -I..\cmake\..\third_party\benchmark\include -I..\third_party\onnx -
Ithird_party\onnx -I..\third_party\foxi -Ithird_party\foxi -I..\caffe2\..\torch\csrc\api -I..\caffe2\..\torch\csrc\api\include -I..\caffe2\aten\src\TH -Icaffe2\aten\src\TH -I..\caffe2\..\torch\..\aten\s
rc -Icaffe2\aten\src -Icaffe2\..\aten\src -Icaffe2\..\aten\src\ATen -I..\caffe2\..\torch\csrc -I..\caffe2\..\torch\..\third_party\miniz-2.0.8 -I..\aten\src\TH -I..\aten\..\third_party\catch\single_inclu
de -I..\aten\src\ATen\.. -Icaffe2\aten\src\ATen -I..\third_party\miniz-2.0.8 -I..\caffe2\core\nomnigraph\include -I..\c10\.. -I..\third_party\cpuinfo\include -I..\third_party\FP16\include -I..\cmake\..\
third_party\googletest\googlemock\include -I..\cmake\..\third_party\googletest\googletest\include -I..\third_party\protobuf\src -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\mkl
\include" -IV:\bin\scripting\python\conda\envs\pytorch\Library\include -I..\cmake\..\third_party\eigen -IV:\bin\scripting\python\conda\envs\pytorch\include -Iv:\bin\scripting\python\conda\envs\pytorch\l
ib\site-packages\numpy\core\include -I..\torch\include -I..\cmake\..\third_party\cub -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\include" -I..\caffe2 /DWIN32 /D_WINDOWS /GR  /w /EHa /MP 
/bigobj -openmp:experimental -DHAVE_AVX_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /Z7 /EHa
 /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -O2 -DCAFFE2_BUILD_MAIN_LIB -DONNX_BUILD_MAIN_LIB -std:c++14 /showIncludes /Focaffe2\CMakeFiles\torch_
cpu.dir\core\blob_serialization.cc.obj /Fdcaffe2\CMakeFiles\torch_cpu.dir\ /FS -c ..\caffe2\core\blob_serialization.cc                                                                                    
V:\src\git\pytorch\src\pytorch\caffe2/core/context_base.h(92): error C2338: CopyToCPU requires fundamental types                                                                                          
..\caffe2\core\blob_serialization.cc(602): note: see reference to function template instantiation 'void at::BaseContext::CopyToCPU<c10::Half>(size_t,const T *,T *)' being compiled                       
        with                                                                                                                                                                                              
        [                                                                                                                                                                                                 
            T=c10::Half                                                                                                                                                                                   
        ]                                                                                                                                                                                                 
[102/995] Building CXX object caffe2\CMakeFiles\torch_cpu.dir\opt\onnxifi_op.cc.obj                                                                                                                       
ninja: build stopped: subcommand failed.                                                                                                                                                                  
Traceback (most recent call last):                                                                                                                                                                        
  File "setup.py", line 755, in <module>                                                                                                                                                                  
    build_deps()                                                                                                                                                                                          
  File "setup.py", line 316, in build_deps                                                                                                                                                                
    cmake=cmake)                                                                                                                                                                                          
  File "v:\src\git\pytorch\src\pytorch\tools\build_pytorch_libs.py", line 62, in build_caffe2                                                                                                             
    cmake.build(my_env)                                                                                                                                                                                   
  File "v:\src\git\pytorch\src\pytorch\tools\setup_helpers\cmake.py", line 337, in build                                                                                                                  
    self.run(build_args, my_env)                                                                                                                                                                          
  File "v:\src\git\pytorch\src\pytorch\tools\setup_helpers\cmake.py", line 141, in run                                                                                                                    
    check_call(command, cwd=self.build_dir, env=env)                                                                                                                                                      
  File "v:\bin\scripting\python\conda\envs\pytorch\lib\subprocess.py", line 347, in check_call                                                                                                            
    raise CalledProcessError(retcode, cmd)                                                                                                                                                                
subprocess.CalledProcessError: Command '['cmake', '--build', '.', '--target', 'install', '--config', 'Release', '--', '-j', '8']' returned non-zero exit status 1.
```                                       `


cc @peterjc123

### PR DESCRIPTION
**Pull request by peterjc123**
Fixes https://github.com/pytorch/pytorch/issues/30932

### CODE REVIEW

File: c10/util/TypeTraits.h [hunk 1]
```diff
@@ -120,5 +120,14 @@ template<template<class> class C>
 struct is_type_condition<C, guts::enable_if_t<std::is_same<bool, guts::remove_cv_t<decltype(C<int>::value)>>::value>> : std::true_type {};
 
 
+/**
+ * is_fundamental<T> is true_type iff the lambda type T is a fundamental type (that is, arithmetic type, void, or nullptr_t).
+ * Example:
+ *  is_fundamental<int> // true
```
Comment (ezyang) [inline c10/util/TypeTraits.h]: We should have an explanation here on why we redefined `is_fundamental` ourselves.

**Review by ezyang — APPROVED** [review]
Just wanting the comment. NB: I won't be able to land this diff for you, please find someone else to land it.

**Comment by mingbowan** [comment]
i'll land this.

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit ea494b1:
* **1/1** broken upstream at merge base 6ab2d1b (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=6ab2d1b1a49de255da11d80772a0dd7ec9e9bae2&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see its [recency history](https://dr.pytorch.org/master-viable-commits.html)):
        * If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
          ```
          git fetch viable/strict
          git rebase --onto viable/strict $(git merge-base origin/master HEAD)
          ```
        * If your commit is older than `viable/strict`:
          ```
          git fetch viable/strict
          git rebase viable/strict
          ```
* **0/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=ea494b1c8a0f7a5632e2828cbd7b341f355c8300).

### 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [pytorch_xla_linux_xenial_py3_6_clang7_build](https://dr.pytorch.org/build-details.html?build_id=73009378)



---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

**Comment by peterjc123** [comment]
@ezyang Can you please take a look at the build error in the fb internal CI?

**Comment by ezyang** [comment]
^ @mingbowan (I don't have access to internal ci results in china)

**Pull request closed without merging**